### DISCUSSION TITLE
Issue with forcing API rules filtering

### DISCUSSION THREAD

**Discussion by speryt** [pocketbase/pocketbase#4708] [Q&A] [answered]
I'm trying to understand API filters behavior, or what I might be doing incorrect.

When I set my collection rule for List/Search as open for **everyone**, everything is working.
When I set it to **admin only** and provide proper credentials in my SvelteKit app
```
export const pb = new Pocketbase('<url>');
await pb.admins.authWithPassword('<email>', '<password>');
```
Also, everything is working fine.

When I changed the login credentials in the app to be for user not admin
`await pb.collection('users').authWithPassword('<email>', '<password>');`
While still keeping the rule to enable List/Search **only for admins**, as expected I am not getting any records. And this is applicable to request performed on my website as well as when I use directly URL generated by request (http://\<url\>/__data.json?page=12&x-sveltekit-trailing-slash=1&x-sveltekit-invalidated=01)

However, the picture changes, when I want to do more advanced rule.
I've added additional select field to my `users` collection - `accountType` with values `regularUser` and `faceless`. I set the value of that field for my user to `faceless` and changed the rule accordingly:
`@request.auth.id != "" && @request.auth.accountType = "faceless"`

What I'd expect now is when I login into my app as this exact user I'll get proper response - which is true, because `accountType` is `faceless`.
What I'm not expecting, is when direct opening of http://\<url\>/__data.json?page=12&x-sveltekit-trailing-slash=1&x-sveltekit-invalidated=01 to get exactly the same response as if filter wouldn't exist at all.
Initially, I've assumed it might be because of some session details stored in my browser, but even when I open incognito window and paste in only this URL, I'm still getting all of the result as if the API was open for everyone. Same behavior when I used this URL in completely different browser.

Is this expected behavior, or what I'm doing wrong?

**Comment by ganigeorgiev** [comment]
It is not clear to which collection you are sending the request, nor what is the expected response.

If you set the rule `@request.auth.id != "" && @request.auth.accountType = "faceless"` in the `users` collection, this rule will be valid only for the specific action of that `users` collection, aka. it grant access to list/view all `users` if the authenticated user has `accountType = faceless`.

If you need further help, please elaborate on your use case and/or export your collections configuration as json (_Admin UI > Settings > Export collections_).

**Reply by speryt** [reply]
Oh no sorry. Maybe I wasn't correct with rule setting part.
All of the rules are set for my `offers` collection, which I'd say is my main collection for the website.

In other words, I have collection `offers`. And I'm doing following call in my app
`const resultList = await pb.collection("offers").getList(page, perPage, options);`

I am expecting that content of this result list will be available only for designated user, that is the `faceless` account.

What I've started to suspect right now (I'm not so fluent in Svelte/SvelteKit so that might be reason of my confusion) is that when I am using URL directly instead of listing records through app I am still effectively going through whole app pipeline. In other words, the URL I provided is not API call to PcocketBase, but rather pseud-API call through my app. If that makes sense.

In such case probably question would be, how can I prevent that? As it is right now I'm afraid this URL might get abused by bad actors at some point. But I think it's more of a SvelteKit question than PocketBase.

**Reply by ganigeorgiev** [reply, marked as answer]
Unfortunately I don't think I'll be able to help you as it still not clear to me how the request is executed and in what context (client-side, server-side, both?).
I suspect that it is either some sort of a cached response (it depends on your SvelteKit configurations and adapter) or if you are using SvelteKit with JS SSR you are probably not initializing a new SDK instance per request, meaning that the node server is using a shared `pb` variable (_see the example in [SSR integration](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#ssr-integration); using a single global SDK instance is safe only when authenticated as admin or in a client-side/browser context_).

The easiest way to rule out PocketBase issue is to send direct request against the PocketBase server with curl, Postman, Insomnia, etc. Alternatively, you can also put the below plain html file in `pb_public/index.html`:

```html
<!DOCTYPE html />
<html>
<head>
    <meta charset="utf-8" />
    <title>Test</title>
</head>
<body>
    <script type="module">
        import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@0.21.1/dist/pocketbase.es.mjs";

        const pb = new PocketBase("http://127.0.0.1:8090");

        // comment the below and open in a new private window to ensure that your rule is applied correctly
        await pb.collection("users").authWithPassword("test@example.com", "1234567890")

        const resultList = await pb.collection("offers").getList(1, 20);

        console.log(resultList)
    </script>
</body>
</html>
```
Navigate to `http://127.0.0.1:8090` and check the browser console.

**Reply by speryt** [reply]
Thank you very much for help.

After reading about SSR, I am certain this might be part of the problem why I'm seeing the behavior I am seeing. Especially since my request is happening on the server side as far as I can tell. My call `const resultList = await pb.collection("offers").getList(page, perPage, options);` is taking place in in `+page.server.ts`

I have also double checked with the html code and indeed when I am logging in as a user see responses and when I have commented this part out I'm not getting any response.

If I may one extra question about SSR integrations. In the part about SvelteKit as I understand it relies on the fact that each user will log in before entering website. But there isn't any (simple) way that would allow any one to get content of the database while accessing through website, but not when using direct URL, right?
